File: app/src/main/java/com/hayatkafeef/game/world/EventLog.java
package com.hayatkafeef.game.world;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Rolling log of meaningful in-game events. Used by the Event Log
 * screen ("ما الذي حدث للتو؟") and by AiSummaryManager to build the
 * end-of-day summary.
 */
public class EventLog {

    public static class Entry {
        public final String timeStr;
        public final String text;
        public final long whenMs;
        public Entry(String t, String s) {
            this.timeStr = t; this.text = s; this.whenMs = System.currentTimeMillis();
        }
    }

    private static final int MAX = 60;
    private final Deque<Entry> entries = new ArrayDeque<>();
    private Entry lastSpoken;

    public synchronized void add(String timeStr, String text) {
        if (text == null) return;
        String t = text.trim();
        if (t.isEmpty()) return;
        if (entries.size() >= MAX) entries.pollFirst();
        Entry e = new Entry(timeStr, t);
        entries.offerLast(e);
        lastSpoken = e;
    }

    public synchronized Entry last() { return lastSpoken; }

    public synchronized List<Entry> recent(int n) {
        List<Entry> list = new ArrayList<>(entries);
        if (list.size() > n) list = list.subList(list.size() - n, list.size());
        return new ArrayList<>(list);
    }

    public synchronized String renderRecent(int n) {
        StringBuilder sb = new StringBuilder();
        for (Entry e : recent(n)) {
            sb.append(e.timeStr).append(" — ").append(e.text).append('\n');
        }
        return sb.toString().trim();
    }

    public synchronized void clear() { entries.clear(); lastSpoken = null; }
}

File: Assets/Scripts/World/EventLog.cs
using System.Collections.Generic;
using System.Text;

namespace BlindLife.World
{
    /// <summary>
    /// Rolling 60-entry log of meaningful in-game events used by
    /// the Event Log overlay and as input to AI summaries.
    /// </summary>
    public class EventLog
    {
        public struct Entry
        {
            public string time;
            public string text;
        }

        const int Max = 60;
        readonly LinkedList<Entry> _entries = new LinkedList<Entry>();
        public Entry? Last { get; private set; }

        public void Add(string time, string text)
        {
            if (string.IsNullOrWhiteSpace(text)) return;
            if (_entries.Count >= Max) _entries.RemoveFirst();
            var e = new Entry { time = time, text = text.Trim() };
            _entries.AddLast(e);
            Last = e;
        }

        public List<Entry> Recent(int n)
        {
            var list = new List<Entry>(_entries);
            if (list.Count > n) list = list.GetRange(list.Count - n, n);
            return list;
        }

        public string Render(int n)
        {
            var sb = new StringBuilder();
            foreach (var e in Recent(n))
            {
                sb.Append(e.time).Append(" — ").Append(e.text).Append('\n');
            }
            return sb.ToString().Trim();
        }
    }
}
